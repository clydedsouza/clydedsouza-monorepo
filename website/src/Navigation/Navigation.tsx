import { Outlet, Link } from "react-router-dom";

function Navigation() {
    return (
      <div>
       <nav>
        <ul>
          <li>
            <Link to="/">Home</Link>
          </li>
          <li>
            <Link to="/projects">Blogs</Link>
          </li>
          <li>
            <Link to="/headline">Healdine</Link>
          </li>
        </ul>
      </nav>

      <Outlet />
      </div>
    );
  }
  
  export default Navigation;