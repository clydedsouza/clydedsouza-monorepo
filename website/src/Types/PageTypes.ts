export enum PageTypes {
  About = 'About',
  Projects = 'Projects',
  Speaking = 'Speaking',
  Teaching = 'Teaching',
  Books = 'Books',
  Highlights = 'Highlights',
  Pinned = 'Pinned',
}
