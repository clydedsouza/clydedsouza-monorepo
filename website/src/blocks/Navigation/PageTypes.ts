export enum PageTypes {
  About = "About",
  Portfolio = "Portfolio",
  Platforms = "Platforms",
  Highlights = "Highlights",
  Cta = "Cta",
  Blog = "Blog",
}
