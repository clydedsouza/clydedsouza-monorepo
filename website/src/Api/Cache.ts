import { PageTypes } from '../Types/PageTypes'
import { IProjectData } from '../Types/ProjectData'
import { getProjectData } from './Core'

const projectDataCacheMap: Map<PageTypes, Promise<IProjectData>> = new Map()

export const getCachedProjectData = (
  pageType: PageTypes
): Promise<IProjectData> => {
  if (!projectDataCacheMap.has(pageType)) {
    const getProjectDataPromise = getProjectData(pageType).catch((error) => {
      projectDataCacheMap.delete(pageType)
      return Promise.reject(error)
    })
    projectDataCacheMap.set(pageType, getProjectDataPromise)
  }

  return projectDataCacheMap.get(pageType)!
}
