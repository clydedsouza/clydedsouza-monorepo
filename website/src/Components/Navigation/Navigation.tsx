import { BrowserRouter, NavLink, Outlet, Route, Routes } from 'react-router-dom'
import AboutMe from '../../Pages/AboutMe/AboutMe'
import Projects from '../../Pages/Projects/Projects'
import { PageTypes } from '../../Types/PageTypes'
import './Navigation.scss'

function Menu() {
  return (
    <>
      <nav>
        <NavLink
          to="/"
          className={({ isActive }) => (isActive ? 'active' : '')}
        >
          {PageTypes.About}
        </NavLink>
        <NavLink
          to={PageTypes.Highlights.toLowerCase()}
          className={({ isActive }) => (isActive ? 'active' : '')}
        >
          {PageTypes.Highlights}
        </NavLink>
        <NavLink
          to={PageTypes.Platforms.toLowerCase()}
          className={({ isActive }) => (isActive ? 'active' : '')}
        >
          {PageTypes.Platforms}
        </NavLink>
      </nav>
      <Outlet />
    </>
  )
}

function Navigation() {
  return (
    <BrowserRouter>
      <Routes>
        <Route path="/" element={<Menu />}>
          <Route index element={<AboutMe />} />
          <Route
            path={PageTypes.Highlights.toLowerCase()}
            element={<Projects name={PageTypes.Highlights} />}
          />
          <Route
            path={PageTypes.Platforms.toLowerCase()}
            element={<Projects name={PageTypes.Platforms} />}
          />
          <Route path="*" element={<AboutMe />} />
        </Route>
      </Routes>
    </BrowserRouter>
  )
}

export default Navigation
