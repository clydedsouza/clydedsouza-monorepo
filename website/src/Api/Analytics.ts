import mixpanel from 'mixpanel-browser'
import { PageTypes } from '../Components/Navigation/PageTypes'
import {
  ILinkClickedProps,
  MIXPANEL_DEV_PROJECT_ID,
  MIXPANEL_PROD_PROJECT_ID,
} from './IAnalytics'

export const initAnalyticsWithSuperProperties = () => {
  const mixpanelProjectId =
    process.env.NODE_ENV === 'production'
      ? MIXPANEL_PROD_PROJECT_ID
      : MIXPANEL_DEV_PROJECT_ID

  try {
    mixpanel.init(mixpanelProjectId)
    mixpanel.register({
      environment: process.env.NODE_ENV,
    })
  } catch {
    undefined
  }
}

export const sendPageViewEvent = (pageType: PageTypes) => {
  try {
    mixpanel.track('Page view', { page: pageType })
  } catch {
    undefined
  }
}

export const sendLinkClickedEvent = (props: ILinkClickedProps) => {
  try {
    mixpanel.track('Link clicked', {
      destination_url: props.link,
      link_type: props.type,
      page_location: props.location,
    })
  } catch {
    undefined
  }
}
