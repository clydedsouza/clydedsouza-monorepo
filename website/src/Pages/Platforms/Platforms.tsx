import { IProject } from '../../Api/IProjectData'
import Cards from '../../Components/Cards/Cards'
import { PageTypes } from '../../Components/Navigation/PageTypes'
import Seo from '../../Components/Seo/Seo'

function Platforms() {
  const sortPlatformsByIdAsc = (projectA: IProject, projectB: IProject) => {
    if (!projectA.date || !projectB.date) return 0
    return projectA.id.localeCompare(projectB.id)
  }

  return (
    <>
      <Seo
        title={PageTypes.Platforms}
        image="https://files.clydedsouza.net/site/ogimage-platforms.png"
        description="Clyde D'Souza is software engineer and author based in Auckland, New Zealand. Feel free to send him a tweet @ClydeDz"
      />
      <Cards pageType={PageTypes.Platforms} sortCards={sortPlatformsByIdAsc} />
    </>
  )
}

export default Platforms
