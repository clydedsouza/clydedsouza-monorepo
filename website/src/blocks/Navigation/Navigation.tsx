"use client";

import { sendNavigationClickedEvent } from "@/api/Analytics";
import Link from "next/link";
import { usePathname } from "next/navigation";
import { useState } from "react";
import "./Navigation.scss";
import { PageTypes } from "./PageTypes";

const getCurrentPage = (pathFromNextJs: string) => {
  const cleanPath = pathFromNextJs.replaceAll("/", "").toLowerCase();

  if (cleanPath === PageTypes.Platforms.toLowerCase()) {
    return PageTypes.Platforms;
  } else if (cleanPath === PageTypes.Portfolio.toLowerCase()) {
    return PageTypes.Portfolio;
  } else {
    return PageTypes.About;
  }
};

function Navigation() {
  const [currentPage, setCurrentPage] = useState(getCurrentPage(usePathname()));

  const onNavigationItemClick = (pageType: PageTypes) => {
    sendNavigationClickedEvent(pageType);
    setCurrentPage(pageType);
  };

  return (
    <>
      <nav>
        <Link
          href="/"
          className={currentPage === PageTypes.About ? "active" : ""}
          onClick={() => onNavigationItemClick(PageTypes.About)}
        >
          {PageTypes.About}
        </Link>
        <Link
          href={PageTypes.Portfolio.toLowerCase()}
          className={currentPage === PageTypes.Portfolio ? "active" : ""}
          onClick={() => onNavigationItemClick(PageTypes.Portfolio)}
        >
          {PageTypes.Portfolio}
        </Link>
        <Link
          href={PageTypes.Platforms.toLowerCase()}
          className={currentPage === PageTypes.Platforms ? "active" : ""}
          onClick={() => onNavigationItemClick(PageTypes.Platforms)}
        >
          {PageTypes.Platforms}
        </Link>
        <Link
          href={"https://blog.clydedsouza.net"}
          target="_blank"
          onClick={() => sendNavigationClickedEvent(PageTypes.Blog)}
        >
          {PageTypes.Blog}
        </Link>
      </nav>
    </>
  );
}

export default Navigation;
