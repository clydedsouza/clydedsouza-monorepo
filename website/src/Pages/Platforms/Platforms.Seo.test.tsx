import {
  render,
  screen,
  waitFor,
  waitForElementToBeRemoved,
} from '@testing-library/react'
import * as apiCacheModule from '../../Api/Cache'
import { PageTypes } from '../../Components/Navigation/PageTypes'
import Platforms from './Platforms'

jest.mock('../../Api/Cache')

describe('Platforms', () => {
  beforeEach(() => {
    jest
      .spyOn(apiCacheModule, 'getCachedProjectData')
      .mockImplementationOnce(() =>
        Promise.resolve({
          app: {},
          data: [],
        })
      )
  })

  it('should render platforms seo', async () => {
    render(<Platforms />)
    await waitForElementToBeRemoved(() => screen.queryByRole('progressbar'))
    await waitFor(() =>
      expect(document.title).toEqual(
        `${PageTypes.Platforms} – Clyde D'Souza – Software Engineer and Author`
      )
    )
    expect(document.head).toMatchSnapshot()
  })
})
