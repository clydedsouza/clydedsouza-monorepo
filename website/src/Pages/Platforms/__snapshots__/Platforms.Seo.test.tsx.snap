// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Platforms should render platforms seo 1`] = `
<head>
  <title>
    Platforms – Clyde D'Souza – Software Engineer and Author
  </title>
  <link
    data-rh="true"
    href="https://clydedsouza.net/#/platforms"
    rel="canonical"
  />
  <meta
    content="Clyde D'Souza is software engineer and author based in Auckland, New Zealand. Feel free to send him a tweet @ClydeDz"
    data-rh="true"
    name="description"
  />
  <meta
    content="https://clydedsouza.net/#/platforms"
    data-rh="true"
    property="og:url"
  />
  <meta
    content="Platforms – Clyde D'Souza – Software Engineer and Author"
    data-rh="true"
    property="og:title"
  />
  <meta
    content="Clyde D'Souza is software engineer and author based in Auckland, New Zealand. Feel free to send him a tweet @ClydeDz"
    data-rh="true"
    property="og:description"
  />
  <meta
    content="https://files.clydedsouza.net/site/ogimage-platforms.png"
    data-rh="true"
    property="og:image"
  />
  <meta
    content="Clyde D'Souza is software engineer and author based in Auckland, New Zealand. Feel free to send him a tweet @ClydeDz"
    data-rh="true"
    property="twitter:card"
  />
  <meta
    content="Platforms – Clyde D'Souza – Software Engineer and Author"
    data-rh="true"
    property="twitter:title"
  />
  <meta
    content="Clyde D'Souza is software engineer and author based in Auckland, New Zealand. Feel free to send him a tweet @ClydeDz"
    data-rh="true"
    property="twitter:description"
  />
  <meta
    content="https://files.clydedsouza.net/site/ogimage-platforms.png"
    data-rh="true"
    property="twitter:image"
  />
</head>
`;
